test/auto_parallel/hybrid_strategy/parallel_api.py

```diff
@@ -201,17 +215,46 @@ def parallel_model(self, layer, optimizer=None):
                 layer, optimizer, self.level
             )
         if self.mp > 1:
-            plan = {
-                "llama.embed_tokens": ColWiseParallel(),
-                "llama.layers.*.self_attn.q_proj": ColWiseParallel(),
-                "llama.layers.*.self_attn.k_proj": ColWiseParallel(),
-                "llama.layers.*.self_attn.v_proj": ColWiseParallel(),
-                "llama.layers.*.self_attn.o_proj": RowWiseParallel(),
-                "llama.layers.*.mlp.gate_proj": ColWiseParallel(),
-                "llama.layers.*.mlp.up_proj": ColWiseParallel(),
-                "llama.layers.*.mlp.down_proj": RowWiseParallel(),
-                "lm_head.weight": ColWiseParallel(),
-            }
+            if not self.sequence_parallel:
+                plan = {
+                    "llama.embed_tokens": ColWiseParallel(),
+                    "llama.layers.*.self_attn.q_proj": ColWiseParallel(),
+                    "llama.layers.*.self_attn.k_proj": ColWiseParallel(),
+                    "llama.layers.*.self_attn.v_proj": ColWiseParallel(),
+                    "llama.layers.*.self_attn.o_proj": RowWiseParallel(),
+                    "llama.layers.*.mlp.gate_proj": ColWiseParallel(),
+                    "llama.layers.*.mlp.up_proj": ColWiseParallel(),
+                    "llama.layers.*.mlp.down_proj": RowWiseParallel(),
+                    "lm_head.weight": ColWiseParallel(),
+                }
+            if self.sequence_parallel:
+                plan = {
+                    "llama.embed_tokens": [
+                        ColWiseParallel(),
+                        PrepareLayerOutput(self.output_transpose),
+                    ],
+                    "llama.layers.*.self_attn.q_proj": [
+                        ColWiseParallel(),
+                        PrepareLayerOutput(self.output_transpose),
+                    ],
+                    "llama.layers.*.self_attn.k_proj": [
+                        ColWiseParallel(),
+                        PrepareLayerOutput(self.output_transpose),
+                    ],
+                    "llama.layers.*.self_attn.v_proj": [
+                        ColWiseParallel(),
+                        PrepareLayerOutput(self.output_transpose),
+                    ],
```

Comment:
according to the principle of sequence parallel, should be `PrepareLayerInput` ?
and should reshard before transpose?